# Issue Sizing Rules

Dev Agent가 바로 작업 가능한 Issue를 만들기 위한 규칙입니다.

## 핵심 원칙

```
1 Issue = 독립적으로 완결성을 가지는 최소 작업단위
```

### 정의

| 속성 | 의미 | 판단 기준 |
|------|------|----------|
| **최소(Atomic)** | 더 쪼개면 의미가 없어짐 | 분할 시 오히려 복잡해지는가? |
| **완결(Self-Contained)** | Issue만 보고 바로 시작 가능 | 외부 참조 없이 작업 가능한가? |
| **독립(Independent)** | 다른 작업 완료를 기다리지 않음 | 선행 작업 없이 시작 가능한가? |

> **30분 Rule**: 위 조건을 만족하면 대체로 30분 이내 완료 가능.
> "30분"은 휴리스틱이지 본질이 아님. 본질은 **"최소 완결 단위"**.

---

## Issue Size 기준

### Good Size (1 Issue)

| 예시 | 예상 시간 |
|------|----------|
| Hero 섹션 타이틀 텍스트 변경 | 5분 |
| CTA 버튼 색상 변경 | 5분 |
| Footer에 이메일 주소 추가 | 10분 |
| 특정 컴포넌트에 props 추가 | 15분 |
| 단일 API 엔드포인트 추가 | 20분 |
| 단일 컴포넌트 생성 | 25분 |

### Bad Size (쪼개야 함)

| 예시 | 문제점 | 분할 방법 |
|------|--------|----------|
| "랜딩 페이지 리뉴얼" | 너무 큼 | Hero, About, Features, CTA 각각 Issue로 |
| "회원 시스템 구축" | 너무 큼 | 가입폼, 로그인, 비밀번호 재설정 각각 |
| "성능 최적화" | 범위 불명확 | 구체적 병목 지점별로 |
| "버그 수정들" | 복수 작업 | 버그 하나당 Issue 하나 |

---

## 분할 기준

### 최소 완결 단위 체크

```
더 쪼갤 수 있는가?
├─ 쪼개면 더 복잡해짐 → ✅ 최소 단위 (유지)
└─ 쪼개도 각각 완결됨 → ❌ 분할 필요

※ 경험적 가이드: 30분 초과 예상 시 분할 검토
```

### Single Responsibility

```
하나의 Issue는 하나의 변경만 담당

❌ "Hero 섹션 수정 + Footer 추가"
✅ Issue #1: "Hero 섹션 수정"
✅ Issue #2: "Footer 추가"
```

### File Count Guide

| 수정 파일 수 | 판단 |
|-------------|------|
| 1-2개 | ✅ 적절 |
| 3-4개 | ⚠️ 분할 검토 |
| 5개 이상 | ❌ 반드시 분할 |

---

## Self-Contained Issue 원칙

### 필수 포함 항목

```markdown
## 이 Issue에서 해야 할 것
[단 하나의 명확한 작업]

## 수정할 파일
[정확한 파일 경로]

## 완료 조건
[확인 가능한 체크리스트]
```

### 금지 사항

```markdown
❌ "자세한 내용은 이 문서를 참고하세요"
❌ "이전 Issue #XX를 먼저 확인하세요"
❌ "프로젝트 가이드라인을 따르세요"
❌ "적절히 판단해서 구현하세요"
```

### 허용 사항

```markdown
✅ 구체적인 파일 경로 명시
✅ 정확한 변경 내용 기술
✅ 확인 가능한 완료 조건
✅ 필요한 코드 스니펫 직접 포함
```

---

## 작업 분할 예시

### Before (Bad)

```
Issue: 랜딩 페이지 개선
- Hero 섹션 새로 만들기
- About 섹션 추가
- Features 섹션 추가
- CTA 버튼 개선
- Footer 연락처 추가
- 반응형 대응
```

### After (Good)

```
Task Group: 랜딩 페이지 개선 (6 Issues)

[Parallel - 동시 가능]
├── Issue #10: Hero 타이틀/서브타이틀 변경
├── Issue #11: About 섹션 컴포넌트 생성
├── Issue #12: Features 섹션 컴포넌트 생성
├── Issue #13: Footer 연락처 추가
│
[Sequential - 순서대로]
├── Issue #14: CTA 버튼 컴포넌트 생성 (먼저)
└── Issue #15: Hero에 CTA 버튼 연결 (Issue #14 이후)
```

---

## PM 책임

PM은 Issue 생성 시 반드시:

1. **최소 단위 체크** - 더 쪼개면 오히려 복잡해지는가?
2. **완결성 체크** - 외부 참조 없이 바로 시작 가능한가?
3. **독립성 체크** - 선행 작업 없이 시작 가능한가?
4. **Single Responsibility 체크** - 하나의 작업만 담고 있는가?
5. **파일 수 체크** - 5개 이상이면 분할 검토
6. **순차/병렬 구분** - Task Group으로 명시

---

## Quick Reference

```
✅ 1 Issue = 독립적 완결성을 가진 최소 작업단위
✅ 더 쪼개면 의미 없어지는 단위
✅ Issue만 보고 바로 작업 시작
✅ 외부 문서 참조 금지
✅ 순차/병렬 명확히 구분

※ 30분은 휴리스틱, 본질은 "최소 완결 단위"
```
